File: design_on_opencv.cpp
#include<iostream>
using namespace std;
#include<opencv2/opencv.hpp>
using namespace cv;

void find_rect( Mat &src, int limit_value)
{
	threshold(src, src, limit_value, 255, THRESH_BINARY);

	int _width = src.cols;
	int _height = src.rows;
	int half_width = 0.5 * _width;
	int half_height = 0.5 * _height;

	int left_x = 0, left_y = 0;
	int right_x = 0, right_y = 0;
	int up_x = 0, up_y = 0;
	int down_x = half_width, down_y = _height;

	int left_flag = 0;
	int right_flag = 0;
	int up_flag = 0;
	int down_flag = 0;
	for (int i = half_width, j = half_height; i > 5; i -= 2)
	{
		int box_mean = (src.at<cv::Vec3b>(j - 1, i - 1)[0] + src.at<cv::Vec3b>(j - 1, i)[0] + src.at<cv::Vec3b>(j - 1, i + 1)[0] +
			src.at<cv::Vec3b>(j, i - 1)[0] + src.at<cv::Vec3b>(j, i)[0] + src.at<cv::Vec3b>(j, i + 1)[0] +
			src.at<cv::Vec3b>(j + 1, i - 1)[0] + src.at<cv::Vec3b>(j + 1, i)[0] + src.at<cv::Vec3b>(j + 1, i + 1)[0]) / 9;
		if ((box_mean > limit_value)&&(!left_flag))
		{
			printf("[%d,%d]", i, j); left_x = i; left_y = j;
			left_flag = 1;
		}		
	}

	for (int i = half_width, j = half_height; i < _width - 5; i += 2)
	{
		int box_mean = (src.at<cv::Vec3b>(j - 1, i - 1)[0] + src.at<cv::Vec3b>(j - 1, i)[0] + src.at<cv::Vec3b>(j - 1, i + 1)[0] +
			src.at<cv::Vec3b>(j, i - 1)[0] + src.at<cv::Vec3b>(j, i)[0] + src.at<cv::Vec3b>(j, i + 1)[0] +
			src.at<cv::Vec3b>(j + 1, i - 1)[0] + src.at<cv::Vec3b>(j + 1, i)[0] + src.at<cv::Vec3b>(j + 1, i + 1)[0]) / 9;
		if ((box_mean > limit_value)&&(!right_flag))
		{
			printf("[%d,%d]", i, j); right_x = i; right_y = j;
			right_flag = 1;
		}
			
	}

	for (int i = half_width, j = half_height; j > 5; j -= 2)
	{
		int box_mean = (src.at<cv::Vec3b>(j - 1, i - 1)[0] + src.at<cv::Vec3b>(j - 1, i)[0] + src.at<cv::Vec3b>(j - 1, i + 1)[0] +
			src.at<cv::Vec3b>(j, i - 1)[0] + src.at<cv::Vec3b>(j, i)[0] + src.at<cv::Vec3b>(j, i + 1)[0] +
			src.at<cv::Vec3b>(j + 1, i - 1)[0] + src.at<cv::Vec3b>(j + 1, i)[0] + src.at<cv::Vec3b>(j + 1, i + 1)[0]) / 9;
		if ((box_mean > limit_value) && (!up_flag))
		{
			printf("[%d,%d]", i, j); up_x = i; up_y = j;
			up_flag = 1;
		}		
	}

	for (int i = half_width, j = half_height; j < _height - 5; j += 2)
	{
		int box_mean = (src.at<cv::Vec3b>(j-1, i-1)[0] + src.at<cv::Vec3b>(j - 1, i)[0] + src.at<cv::Vec3b>(j - 1, i + 1)[0] +
			src.at<cv::Vec3b>(j, i-1)[0] + src.at<cv::Vec3b>(j, i)[0] + src.at<cv::Vec3b>(j, i + 1)[0] +
			src.at<cv::Vec3b>(j + 1, i - 1)[0] + src.at<cv::Vec3b>(j + 1, i)[0] + src.at<cv::Vec3b>(j + 1, i + 1)[0])/9;
		if ((box_mean > limit_value)&&(!down_flag))
		{
			printf("[%d,%d]", i, j); down_x = i; down_y = j;
			down_flag = 1;
		}
	}

	rectangle(src, Rect(left_x, up_y, right_x - left_x, down_y - up_y), Scalar(0, 255, 0), 3);
}




int main()
{
	Mat src = imread("demo.jpg");

	find_rect(src, 100);

	imshow("0",src);
	waitKey(0);

	return 0;
}

File: design_on_opencv.py
# -*- coding: utf-8 -*-
import cv2 as cv
def find_rect(src,gray):
    up_flag = 1
    down_flag = 1
    left_flag = 1
    right_flag = 1
    rect=[0,0,0,0]
    ret,thresh=cv.threshold(gray,100,255,cv.THRESH_BINARY)
    height = gray.shape[0]
    width = gray.shape[1]
    
    for row in range(int(0.5*height),0,-1):
        col = int(0.5*width)
        pv = thresh[row, col]
        if((pv>150)&(up_flag)):
            up_flag=0
            rect[1]=row
            cv.circle(src,(col,row),3,(0,0,213),-1)
            
    for row in range(int(0.5*height),height):
        col = int(0.5*width)
        pv = thresh[row, col]
        if((pv>150)&(down_flag)):
            down_flag=0
            rect[3]=row
            cv.circle(src,(col,row),3,(0,0,213),-1)
        
    for col in range(int(0.5*width),0,-1):
        row = int(0.5*height)
        pv = thresh[row, col]
        if((pv>150)&(left_flag)):
            left_flag=0
            rect[0]=col
            cv.circle(src,(col,row),3,(0,0,213),-1)
            
    for col in range(int(0.5*width),width):
        row = int(0.5*height)
        pv = thresh[row, col]
        if((pv>150)&(right_flag)):
            right_flag=0
            rect[2]=col
            cv.circle(src,(col,row),3,(0,0,213),-1)
    for col in range(rect[0]+5,rect[2]-5,2):
        for row in range(rect[1]+5,rect[3]-5,2):
            pv1 = [thresh[row-1, col-1],thresh[row-1, col],thresh[row-1, col+1]]
            pv2 = [thresh[row  , col-1],thresh[row  , col],thresh[row  , col+1]]
            pv3 = [thresh[row+1, col-1],thresh[row+1, col],thresh[row+1, col+1]]
            aver =  (int(pv1[0])+int(pv1[1])+int(pv1[2]))/9
            aver += (int(pv2[0])+int(pv2[1])+int(pv2[2]))/9
            aver += (int(pv3[0])+int(pv3[1])+int(pv3[2]))/9
            if(aver > 240):
                cv.circle(src,(col,row),3,(0,0,213),-1)
        cv.rectangle(src, (rect[0],rect[1]), (rect[2],rect[3]), (255,0,0),3)
    cv.imshow("bin", thresh)
    cv.imshow("src", src)
    
    
image = cv.imread("demo.jpg")
GrayImage=cv.cvtColor(image,cv.COLOR_BGR2GRAY)
find_rect(image,GrayImage)


cv.waitKey(0)
cv.destroyAllWindows()
